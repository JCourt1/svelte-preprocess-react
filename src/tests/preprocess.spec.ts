import { describe, expect, it } from "vitest";
import { preprocess } from "svelte/compiler";
import sveltePreprocess from "svelte-preprocess";
import { readFile } from "fs/promises";
import { dirname, resolve } from "path";
import preprocessReact from "../lib/preprocess";

describe("svelte-preprocess-react", () => {
  it("should process <react:component> tags", async () => {
    const filename = resolveFilename("./fixtures/Container.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });
  it("should process <react:component> tags", async () => {
    const filename = resolveFilename("./fixtures/Multiple.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });

  it("should fail on bindings", async () => {
    const filename = resolveFilename("./fixtures/Binding.svelte");
    const src = await readFile(filename, "utf8");
    let failed: boolean;
    try {
      await preprocess(src, preprocessReact(), { filename });
      failed = false;
    } catch (err: any) {
      expect(err.message).toMatchSnapshot();
      failed = true;
    }
    expect(failed).toBe(true);
  });

  it("should fail on slots (for now)", async () => {
    const filename = resolveFilename("./fixtures/Slots.svelte");
    const src = await readFile(filename, "utf8");
    let failed: boolean;
    try {
      await preprocess(src, preprocessReact(), { filename });
      failed = false;
    } catch (err: any) {
      expect(err.message).toBe(
        "Nested components are not (yet) supported in svelte-preprocess-react"
      );
      failed = true;
    }
    expect(failed).toBe(true);
  });

  it("should inject a script tag", async () => {
    const filename = resolveFilename("./fixtures/NoScript.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toContain("<script>");
    expect(output.code).toMatchSnapshot();
  });

  it("should support typescript when using preprocess", async () => {
    const filename = resolveFilename("./fixtures/Typescript.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(
      src,
      preprocessReact({ preprocess: sveltePreprocess() }),
      { filename }
    );
    expect(output.code).toMatchSnapshot();
  });
});

const base = dirname(import.meta.url).replace("file://", "");
function resolveFilename(filename: string) {
  return resolve(base, filename);
}
