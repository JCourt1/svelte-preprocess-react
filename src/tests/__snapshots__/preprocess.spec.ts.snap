// Vitest Snapshot v1

exports[`svelte-preprocess-react > should fail on bindings 1`] = `"'count' is not a valid binding"`;

exports[`svelte-preprocess-react > should inject a script tag 1`] = `
"<script>
import sveltifyReact from \\"svelte-preprocess-react/sveltifyReact18\\";

const React$Counter = sveltifyReact(Counter);
</script>

<!-- Counter could be a global variable -->
<React$Counter />
"
`;

exports[`svelte-preprocess-react > should process <react:component> tags 1`] = `
"<script>import sveltifyReact from \\"svelte-preprocess-react/sveltifyReact18\\";
  // @ts-nocheck
  import Counter from \\"./Counter\\";

  let count = 1;
const React$Counter = sveltifyReact(Counter);</script>

<React$Counter
  {count}
  onCount={(next) => {
    count = next;
  }}
/>
<React$Counter count={2} onCount={() => {}} />
"
`;

exports[`svelte-preprocess-react > should support typescript when using preprocess 1`] = `
"<script lang=\\"ts\\">export let title;
</script>

<h1>{title}</h1>
"
`;
